Create linalg.matrix_exp

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62734
* #62716
* __->__ #62715

Fixes https://github.com/pytorch/pytorch/issues/61648

Differential Revision: [D31641698](https://our.internmc.facebook.com/intern/diff/D31641698)

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

lezcano committed:
Create linalg.matrix_exp

Fixes https://github.com/pytorch/pytorch/issues/61648

[ghstack-poisoned]

lezcano committed:
Update on "Create linalg.matrix_exp"


Fixes https://github.com/pytorch/pytorch/issues/61648

[ghstack-poisoned]

nikitaved commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2052,26 +2052,30 @@ Tensor backward_analytic_function_of_a_matrix(
-              "matrix_exp(", a.scalar_type(), "{", a.sizes(), "}): expected a tensor "
-              "of squared matrices");
+Tensor linalg_matrix_exp(const Tensor& a) {
+  squareCheckInputs(a);
+  TORCH_CHECK((at::isFloatingType(a.scalar_type()) || at::isComplexType(a.scalar_type())),
+              "Expected a floating point or complex tensor as input. Got: ", a.scalar_type());
```
What if we introduce additional parameter to squareCheckInputs indicating the name of the function it fails in if that happens? Same goes for other checks. Will also make the errors coming from `torch.linalg` a touch easier to parse and identify.

Lezcano replied:
I think that's a great idea. I'll do that in a bit.

Lezcano replied:
Done! Could you review it please? https://github.com/pytorch/pytorch/pull/62734

mruberry replied:
The dtype error message should also start with the name of the operator, like "torch.linalg.matrix_exp: Expected..."

Lezcano replied:
This is done in the third PR of this stack.

IvanYashchuk commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2052,26 +2052,30 @@ Tensor backward_analytic_function_of_a_matrix(
+    return a.clone();
+  } else if (n == 1) {
     return a.exp();
+  } else{
```
```suggestion
  } else {
```

IvanYashchuk commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2052,26 +2052,30 @@ Tensor backward_analytic_function_of_a_matrix(
+}
 
-  return mexp(a);
+// Alias
```
See `# Note [Adding an alias]` in `native_functions.yaml`

IvanYashchuk reviewed (changes requested):
Looks great! There are some additional steps needed for an aliased function. See `# Note [Adding an alias]` in `native_functions.yaml`. Steps 3 and 6 are missing from this PR.

lezcano committed:
Update on "Create linalg.matrix_exp"


Fixes https://github.com/pytorch/pytorch/issues/61648

[ghstack-poisoned]

Lezcano commented:
@IvanYashchuk @nikitaved Addressed your comments

lezcano committed:
Update on "Create linalg.matrix_exp"


Fixes https://github.com/pytorch/pytorch/issues/61648

[ghstack-poisoned]

nikitaved reviewed (approved):
LGTM! Thanks, Mario!

lezcano committed:
Update on "Create linalg.matrix_exp"


Fixes https://github.com/pytorch/pytorch/issues/61648

[ghstack-poisoned]

IvanYashchuk reviewed (approved):
Thanks, @Lezcano!

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -10223,6 +10220,14 @@
+  dispatch:
+    CPU, CUDA: linalg_matrix_exp
+
+- func: linalg_matrix_exp_backward(Tensor self, Tensor grad) -> Tensor
```
Why is the backward exposed in native_functions.yaml?

Lezcano replied:
It was there already, this is just renaming and moving around. This is moved to `FunctionsManual.cpp`  in the second PR of this stack.

mruberry replied:
I think matrix_exp_backward is actually used. Can we do the following:

- leave matrix_exp_backward as-is
- not expose a linalg_matrix_exp_backward

lezcano committed:
Update on "Create linalg.matrix_exp"


Fixes https://github.com/pytorch/pytorch/issues/61648

[ghstack-poisoned]

lezcano committed:
Update on "Create linalg.matrix_exp"


Fixes https://github.com/pytorch/pytorch/issues/61648

[ghstack-poisoned]

lezcano committed:
Update on "Create linalg.matrix_exp"


Fixes https://github.com/pytorch/pytorch/issues/61648

[ghstack-poisoned]

iseeyuan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -10254,6 +10253,12 @@
   dispatch:
     CPU, CUDA: linalg_lstsq_out
 
+- func: linalg_matrix_exp(Tensor self) -> Tensor
```
Adding a new operator should be fine. There were a break last time, because some users are still using old selective build method, which is not guarded by CI. This kind of update should be captured now with internal CI and the operator list should be updated in the internal diff. cc @dhruvbird to confirm.

iseeyuan replied:
OK. Confirmed that the old way of selective build is still being used. So removing an operator could break some existing users. We are trying to prioritize the deprecation of old selective build.

cccclai reviewed (approved):
From pytorch edge side, prod models look good

mruberry reviewed (approved):
Thanks for the updates @Lezcano and your review, @cccclai, I think we're good to go!

Lezcano commented:
The day has come!!

lezcano committed:
Update on "Create linalg.matrix_exp"


Fixes https://github.com/pytorch/pytorch/issues/61648

Differential Revision: [D31641698](https://our.internmc.facebook.com/intern/diff/D31641698)

[ghstack-poisoned]
